### ISSUE TITLE
[UX] Make it clear in the manage blocks UI that title-only custom blocks without content will not be rendered.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2640]
STR:

1. Add a new layout with `node/%` as path/context and preferably choose a template with a sidebar.
2. Also add a visibility condition and set it to "content type = post" for example.
3. Add a custom block to the layout (preferably on the sidebar); only title; no content.
4. Create a piece of content of the type you've selected in step 2 above -> the sidebar is rendered (noticed as blank space on the side of the page content), but no custom block 👎 
5. Edit the layout, edit the custom block from step 3 above and this time add some text to its content. Save the block and the layout.
6. Visit the post page again -> block is there now 👍 

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/1835 (makes it so that custom blocks are rendered even if they only have a title and no content)

**Comment by jenlampton** [comment]
This is by design.

Here's a use case where it's important that the block disappear: You have a sidebar item "Recent comments", but there are no recent comments. Having the block heading show up with nothing below it would make the website seem broken. People will look at that and think "Where are the recent comments?"

**Comment by klonos** [comment]
In the specific use case that you mention, I see that as content/configuration issue. I assume that the "Recent comments" block is a view, so if there are no comments, there is also a setting to show a message like "No recent comments". So, there is content in that case.

The specific example aside, what you say is a valid use case. It does not mean that an empty custom block should not be rendered if it has a title though. That is a valid use case too. Otherwise, the site builder would thing it's a bug: "Why isn't it being rendered?".

I have found the code with the logic that renders blocks or not. Checking if there is a way to account for both use cases...

**Comment by klonos** [comment]
...although I must admit that what I was trying to achieve was not so "standard" and could probably be achieved in another way:

I wanted to quickly add some sort of sidebar region title. So, since regions do not currently have that as an option, I would have to do it via css. But then I though that perhaps use a custom block with just a title. This would allow me to click my way through the admin UI without touching any css + would give me the flexibility of using block visibility conditions. It was a WTF moment when I saw that the block would not be rendered, because I had no idea that this is by design.

So, bottom line is this I think:

1. With dynamically rendered blocks it makes sense to not render them at all if there is no content.
2. If the intention is to display the title of dynamically rendered blocks even when there is no content, then site builders should make sure that there is some "empty text message" shown (like with view and views-provided blocks). In that case, the empty text message acts as content and the block would be rendered.
3. With static content blocks like custom blocks, having them with only title and then not rendering that block is plain silly and unexpected behavior. If the site builder would want the block to not be rendered, they'd just remove it.

So how about we unconditionally render the block if it is a custom block (even if it only has a title and no content) and we let the default behavior of not rendering blocks if they have no content apply to the rest of the block types?

**Comment by klonos** [comment]
Something like that perhaps?

**Comment by jenlampton** [comment]
Semantically, the title is a heading for what comes below, in the block.
There's no use case for showing a label for a thing if there's no thing
there to label.

I do understand this is a confusing user experience, especially for custom
blocks, but I dont think we should support showing a title when there's not
a case for only having a title (except maybe testing, like you are doing
now)

Why don't we fix the user experience instead? Let's add an admin
description when there is a title only, letting the user know the block
will not be displayed. Something in g like "this block will not be
displayed without content".

**Comment by klonos** [comment]
> ...I dont think we should support showing a title when there's not a case for only having a title (except maybe testing, like you are doing now)

I wasn't just testing; I was building a site and wanted a quick'n'dirty (site builder, non-coder) way to add a title to a sidebar without touching any code:

>I wanted to quickly add some sort of sidebar region title. So, since regions do not currently have that as an option, I would have to do it via css. But then I though that perhaps use a custom block with just a title. This would allow me to click my way through the admin UI without touching any css...

**Comment by klonos** [comment]
>Why don't we fix the user experience instead? Let's add an admin description when there is a title only, letting the user know the block will not be displayed. Something in g like "this block will not be displayed without content".

I think that in that case, the "proper" way to "fix" this would be to make the content field required and have a validation error thrown (instead of a description text) when no content it entered. I mean, what's the point in adding a title-only custom block when it's not going to be rendered anyway? So:

- Custom blocks with both title and content: business as usual
- Custom blocks with no title; just content: allow
- Custom blocks with just a title and no content: do not allow

**Comment by jenlampton** [comment]
That would work too, but...

One thing I really hate about building menus is the validation errors that
happen when there are no paths set up yet. I would much rather be able to
actually build my links, even if they didn't show up. (My current
work-around is to link every menu item to <front> to avoid the validation
errors.)

I wonder if preventing people from building things (via validation error)
is going to be creating unnecessary frustration in this interface too..

**Comment by klonos** [comment]
Are you suggesting that one possible use case is that people are using "blank" custom blocks (only title) as placeholder blocks until they figure out what content to put in them? Then yes, I honestly have not thought of that, but it does make sense. Then again, doesn't that make it annoying to have the current behavior of not rendering these placeholder blocks until they have some (any) content first?

**Comment by jenlampton** [comment]
There are two things we want to avoid:

1) the system seeming broken when people don't understand how it works
2) the system preventing people from doing something that's part of a usual site-building process (like adding links to menu items that don't exist yet)

I'm not sure #2 really applies here, but I'd hate to create another menu-link scenario by using validation where we could have solved #1 differently.

I'm on the fence about whether we should show things in a half-built state, but I'm leaning towards no (for both menu links and blocks) as long as we can show something on the admin side that explains why they are not displaying on the front end. We do this already with deleted custom blocks, broken/missing handlers, etc.

**Comment by jlfranklin** [comment]
If I am building a site, need a block, but don't have the content for the block, I'll put in some filler text, "Lorem Ipsum" or "Placeholder for XYZ."  In @klonos's case, I'd create a block with no title and content wrapped in an H2 or H3 tag, or a block with a title and an empty DIV or &nbsp; for the content.

For UX/DX purposes, we can make the body a required field in the block editor.  There really isn't any point to a content-free block.

**Comment by Graham-72** [comment]
> I'd create a block with no title and content wrapped in an H2 or H3 tag

I have done exactly this where I needed to have a common heading for a group of two or three blocks, each with its own heading (i.e. subheading). This did seem to be a case where I wanted a block that was just a heading but I'm happy to do this by using content styled as a heading, with no block heading.

**Comment by klonos** [comment]
@jlfranklin yes, dummy "lorem ipsum" is what **I** do as well. And yes, knowing my way around Backdrop/Drupal (been using it for many years), **I**'ve figured that **I** can use H2H3 in the body of the custom block **as a workaround**. This is not about you, Jen, Graham or myself though. The question is whether a novice user would consider this a feature or a bug:

- They create a custom block with only a title.
- When they save the block, there is no validation error, so they assume all is good.
- They see the block in their layout, so they save the layout and once again assume that all is good.
- They switch to the frontend and the block is not there -> WTF!
- They double check the layout and the block is still there -> WTF!

So, I've filed this issue here because we are not meeting the user expectations of how things should work in the UI.

PS (warning: ranting involved): this case reminded me of what happened to my brother while using Drupal to finish one of our College assessments during the weekend. He is not a Drupal/CMS expert; he only has bits and pieces of info of how things work from my rants when we are having discussions about how our days were. So, we were working side by side on our laptops trying to hide some fields from the display of content. I told him to navigate to Structure -> Content types -> whatever -> Manage display and he should figure it out. Having seen the option "hidden" in the dropdown menu for the label of field, he selected that and saved the form. He then said to me "I've hidden it, but it is still shown here.". I said: "That's because you've hidden the field *label*; not the field itself. You need to drag the entire field below the "hidden" row.". I said that as if it was the most obvious thing in the world, and you should then see my brothers facial expression. It's the same as what we're saying here: "Of course the **entire** block will not be rendered; it has only title and no body." and we are saying this as if it's the most obvious thing in the world.

My point is that yes, the check of whether a block has content before rendering is there for a good reason: to not render **dynamically generated** blocks when they have no content. But with custom blocks if the user **intentionally** leaves them without any content, they expect it to be rendered no matter what.

How I envision that we could alternatively solve this UX issue is by having some red text in place of the block description on the manage blocks UI, clearly indicating that the block will not be rendered as is because the body field is empty. Something like this:

![backdrop-broken_block](https://cloud.githubusercontent.com/assets/2423362/19889758/e6bc8a64-a08a-11e6-8bb0-77042dc6c32a.png)

That would solve the use case of false expectations with regards to the block behavior, since that way there would be clear indication of why the block will not be rendered. It does not solve the use case of people actually wanting to render a block with only a title.

As for the suggestion to be using H2/H3 in the body, yes but... the intention was to place a quick and dirty header, *without touching code* and with as little tweaking as possible. The default out of the box is to allow only H3/H4/H5 headers for the body of custom blocks (default setting for the filtered HTML text format), while the rest of the blocks get H2 headers in their titles. Here's how this would look if an H2/H3 header is used within the block content/body:

![image](https://cloud.githubusercontent.com/assets/2423362/25148011/d18124aa-24bc-11e7-8831-ae2021a1aa91.png)

That would mean requiring to adjust the CSS in order to match the style of block H2 and H3 headers, while the intention was to not touch code. Of course the user could switch to full HTML for the block body and then "manually" change the H3 to an H2, but again: no coding. Of course the user could tweak the allowed HTML tags to include H2 under `/admin/config/content/formats/filtered_html`, but a) we said minimal tweaking and b) this could end up messy as blocks with titles could potentially end up with 2 nested H2 tags (one in the title and one in the content).

So are there other ways to achieve adding a header/title? Yes.
Are they as easy and straight-forward to Backdrop/Drupal novices as adding a block title? Nope.
Do we allow them to use the easy way: Yes.
Does it actually work? Nope.

Lets fix it.

**Comment by docwilmot** [comment]
I believe it is standard that if blocks dont have content they dont display, and I think this is a good thing to maintain. But I also agree that the expectation might be that if there is no error, all should work. The idea of having text explaining this, as in the broken block type message is the best option then.

I dont agree with changing things just because a person may want to see what block headers look like. Any gibberish text or even a dot in the body field would suffice as content and force it to render. We dont have to recode Backdrop just for this unlikely possibility.

**Comment by klonos** [comment]
> The idea of having text explaining this, as in the broken block type message is the best option then.

Yes, at the very least that.

**Comment by klonos** [comment]
> I dont agree with changing things just because a person may want to see what block headers look like. ... We dont have to recode Backdrop just for this unlikely possibility.

It was not just to see; it was part of the design and the user was trying to achieve that with minimum effort. If this is seen as abusing the way blocks are meant to work, then fair enough.

**Comment by klonos** [comment]
...to tell you the truth, I would actually prefer it if there was a way to add a "configure" option to regions and have the option to set a title for the entire region.

...or have the block container feature so that one could configure the "parent" block with just a title. Since that container block would have content (its "child" blocks), it would be rendered with its title serving as what was the goal in my use case.

**Comment by docwilmot** [comment]
Was there consensus on this? The PR makes an exception for custom blocks to allow them to show without content. I think this is breaking a standard expectation and needs more input.

The alternative, to have an error message on the add block page (+/- a set message warning) isn't in this PR.

**Comment by jenlampton** [comment]
I also disagree with the approach in the current PR. I think all blocks should behave the same way. If someone wants a H2 tag and nothing else in a block (which is semantically bad form, as well as an edgecase),  the H2 can go in the block content area - yes, as workaround. I much prefer the solution @klonos mentioned that will both educate users as to how (all) blocks work, and solve the UX problem of something appearing to work on the back-end, but not be present on the front-end:

> How I envision that we could alternatively solve this UX issue is by having some red text in place of the block description on the manage blocks UI, clearly indicating that the block will not be rendered as is because the body field is empty.

**Comment by klonos** [comment]
Yes, that PR was filed before this issue was re-titled and re-purposed.

**Comment by daggerhart** [comment]
@klonos I've done similar things before and tend to just use input format Full HTML w/ an html comment as the content: `<!-- placeholder -->`

The only drawback to that I could see is that the content still takes up space, so it if has a lot of padding or margin or something, it would be obvious. 

If that is a problem for your case, you could solve it 2 ways: 

1. Select "Dynamic" for your block style, and give the content area No wrapper, that should get rid of your padding/margins/style.  

Or 2, give the block some custom class like `block-hide-content`, and then in your css add `.block-hide-content { display: none; }`

**Comment by jenlampton** [comment]
I'm going to remove the milestone from this issue since there's no PR yet and we are releasing 1.6.4 today. However, any issue labeled as a "bug" can be included in the next bug-fix release, and as soon as we have a PR for review, the milestone can be added back.

**Comment by LeeteqXV** [comment]
Maybe the two workarounds are acceptable. We will hopefully get a "custom CSS" GUI feature in place "soon" (other issue), which makes it easy enough. The default should be not showing blocks that does not render content. so that we may have consistency across all blocks.

However, I have one particular (repeating) situation in mind that I do not see as an "edge case":

I find it very practical to render empty, "coming" blocks when working on non-public/dev sites, or internal sites that does not have the concern of confusing visitors. I like to prepare content to be placed here and there and have them display only with titles, awaiting content. In the example of 0 comments, that falls into the category of "views blocks". On a coming community with only the structure in place, I would like to see those empty blocks...

So I wonder if we could add a setting like the "override title", to "override hiding empty block, so we can manually make exceptions on suitable sites?

**Comment by klonos** [comment]
FYI, I have filed an initial PR in #2647 that allows setting region titles.

**Comment by BWPanda** [comment]
Unless I've missed something, the only reason I can see given for not allowing title-only blocks is that it's not semantically correct. Do we want to force users to be semantic? Are there other places we do this? If I recall, there's a setting in views for hiding empty blocks, meaning blocks are displayed by default when there's no content, unless you specifically change this setting. Also HTML defines semantic code/recommendations, but it doesn't enforce them (e.g. when to use Hn tags, not putting block elements inside inline ones, etc.). Semantics are up to the user to use or ignore. I therefore think we should let users create title-only blocks and render them as they'd expect.

**Comment by docwilmot** [comment]
> and render them as they'd expect.

the whole point here is that theyd probably expect them not to show up at all, because thats how all other blocks work (didnt know Views blocks were exceptions). this PR was about setting a new precedent, which might have unexpected consequences. (people with empty blocks for whatever reason suddenly showing up in the front page, for example)

**Comment by klonos** [comment]
> people with empty blocks for whatever reason suddenly showing up in the front page, for example

As a compromise, how about we add a warning in the layout, saying something along these lines:

> No content has been added in the custom block _xyz_. This block will not be rendered until it has some content.

**Comment by jenlampton** [comment]
Alternate text proposal:

> No content has been added in the custom block xyz. This block will not be displayed.
